std/range/package.d

```diff
@@ -315,14 +315,20 @@ if (isBidirectionalRange!(Unqual!Range))
                 {
                     import std.algorithm.mutation : move;
 
-                    source.back = move(val);
+                    static if (is(typeof({source.back = move(val);})))
+                        source.back = move(val);
+                    else
+                        source.back = val;
```

Comment:
I think this block of code should probably be a utility function or the condition check should become a utility template, since using `typeof` to see if something compiles is essentially like `__traits(compiles`, which comes with a bunch of downsides and is what I would consider only as a last resort and should certainly not be copy pasted a bunch of times across the code base.

I'm not sure if the condition should check `source.back = move(val)` or `typeof(val) x = move(val)` or only `move(val)` - probably the least to make the tests pass is a good idea

Reply:
`is(typeof(...))` is a commonly used construct, and I don't recall running into any issues with it. What about it do you consider problematic in comparison to `__traits(compiles, ...)`?

As for the exact check within whichever construct is used, I'm pretty sure that it has to check the whole expression, because it could be affected both by the types involved and the exact signature of whichever function the move is going to.

Reply:
since it's basically the same, it shares the same issues as `__traits(compiles, ...)`, where you can easily shoot yourself in the foot whenever the compile-checked code changes in any way, imo a code smell that should only be used sparingly. Especially when it's somewhere inside Phobos changing how code behaves this can lead to linker errors and unintuitive behavior or silently stopping to call certain functions with templated code. (which is almost all the opAssign code)

However if it's fine to use it in Phobos, don't let my comment prevent you from doing so. It's just that I added the `move` there and don't think we should need to fight the language with a bunch of boilerplate for a move assignment. It kind of feels like this might even be good as some kind of compiler fix or enhancement allowing this stuff somehow, but I don't really see how or if that would even work.

This also looks like someone might look at those parts in phobos in the future and think "wtf why is this even needed"

I think the real underlying issue is how can there be a type that is not movable but assignable?

Reply:
Yeah. It's a really weird edge case, though once we have move constructors, it may happen more, since then presumably it will be possible to have assignment while disabling the move constructor - which would still be weird but arguably not as weird as having a const opAssign.

Fundamentally here though, what we need to check is whether a particular piece of code compiles. It's what `hasAssignableElements` and similar traits are doing as well. It's always nice when we can encapsulate that cleanly in a trait, though I'm not sure how that could be done here, since both what's being moved to and what's being moved matters, and the left-hand side differs with each of these.

And I certainly have no problem if someone can come up with a cleaner way to do these checks, but at least these work, and even if someone wonders why it's needed, since there are unit tests that will fail if it's removed, and those tests reference a bug report, anyone who attempts it will be able to find out why it's needed.

That being said, part of me thinks that the language should just require that if assignment works, moving works, since the more stray abilities that we have to test before using in generic code, the more problematic it gets, but I don't know (though obviously, the Weka folks wouldn't be happy with that, since it would break what they're doing).

Reply:
`hasMobileElements` is similar to what we'd need, but that tests whether the elements within the range can be moved, not whether another value can be moved onto the elements of the range.

Reply:
Okay. I updated it to use `forward`, which gets rid of all of the new compilability checks.

I was going to also add some CTFE tests, but I ran into https://issues.dlang.org/show_bug.cgi?id=24527 which gets in the way of verifying that `opAssign` works correctly during CTFE, since in at least some cases, it actually doesn't even if your code is right. But in principle, this code will assign during CTFE rather than try to move.